src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -435,144 +538,136 @@ public void syncOnEachOperation(boolean syncOnEachOperation) {
-                if (translogStream instanceof LegacyTranslogStream) {
-                    input.readInt(); // ignored opSize
+        synchronized void onNewTranslog(FsChannelReader oldCurrent, FsChannelReader newCurrent) throws IOException {
+            // even though the close method removes this view from outstandingViews, there is no synchronisation in place
+            // between that operation and an ongoing addition of a new translog, already having an iterator.
+            // As such, this method can be called despite of the fact that we are closed. We need to check and ignore.
+            if (closed) {
+                // we have to close the new references created for as as we will not hold them
+                IOUtils.close(oldCurrent, newCurrent);
+                return;
+            }
+            FsChannelReader reader = orderedTranslogs.remove(orderedTranslogs.size() - 1);
+            reader.close();
+            orderedTranslogs.add(oldCurrent);
+            orderedTranslogs.add(newCurrent);
+        }
+
+        @Override
+        public synchronized long minTranslogId() {
+            return orderedTranslogs.get(0).translogId();
+        }
+
+        @Override
+        public synchronized int estimatedTotalOperations() {
+            int ops = 0;
+            for (FsChannelReader translog : orderedTranslogs) {
+                int tops = translog.totalOperations();
+                if (tops == FsChannelReader.UNKNOWN) {
+                    return -1;
                 }
-                return translogStream.read(input);
-            } catch (TruncatedTranslogException | EOFException e) {
-                // ignore, not properly written the last op
-                logger.trace("ignoring translog EOF exception, the last operation was not properly written", e);
-                return null;
-            } catch (IOException e) {
-                // ignore, not properly written last op
-                logger.trace("ignoring translog IO exception, the last operation was not properly written", e);
-                return null;
+                ops += tops;
+            }
+            return ops;
+        }
+
+        @Override
+        public synchronized long sizeInBytes() {
+            long size = 0;
+            for (FsChannelReader translog : orderedTranslogs) {
+                size += translog.sizeInBytes();
             }
+            return size;
+        }
+
+        public synchronized Snapshot snapshot() {
+            return createdSnapshot(orderedTranslogs.toArray(new FsChannelReader[orderedTranslogs.size()]));
         }
 
+
         @Override
         public void close() throws ElasticsearchException {
+            List<FsChannelReader> toClose = null;
             try {
-                input.close();
-            } catch (IOException ex) {
-                throw new ElasticsearchException("failed to close stream input", ex);
+                synchronized (this) {
+                    if (closed) {
```

Comment:
can we do `if (closed == false)` or use an AtomicBoolean and use `if (closed.compareAndSet(false, true)) {`